src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -0,0 +1,365 @@
+
+        this.min = min;
+        this.current = min;
+        this.max = max;
+        this.count = count;
+    }
+
+    public final PrimaryKey getMinimum()
+    {
+        return min;
+    }
+
+    public final PrimaryKey getCurrent()
+    {
+        return current;
+    }
+
+    public final PrimaryKey getMaximum()
+    {
+        return max;
+    }
+
+    public final long getCount()
+    {
+        return count;
+    }
+
+    /**
+     * When called, this iterators current position should
+     * be skipped forwards until finding either:
+     *   1) an element equal to or bigger than next
+     *   2) the end of the iterator
+     *
+     * @param nextToken value to skip the iterator forward until matching
+     *
+     * @return The next current token after the skip was performed
+     */
+    public final PrimaryKey skipTo(PrimaryKey nextToken)
+    {
+        if (min == null || max == null)
+            return endOfData();
+
+        // In the case of deferred iterators the current value may not accurately
+        // reflect the next value so we need to check that as well
+        if (current.compareTo(nextToken) >= 0)
+        {
+            next = next == null ? recomputeNext() : next;
+            if (next == null)
+                return endOfData();
+            else if (next.compareTo(nextToken) >= 0)
+                return next;
+        }
+
+        if (max.compareTo(nextToken) < 0)
+            return endOfData();
+
+        performSkipTo(nextToken);
+        return recomputeNext();
+    }
+
+    protected abstract void performSkipTo(PrimaryKey nextToken);
+
+    protected PrimaryKey recomputeNext()
+    {
+        return tryToComputeNext() ? peek() : endOfData();
+    }
+
+    protected boolean tryToComputeNext()
+    {
+        boolean hasNext = super.tryToComputeNext();
+        current = hasNext ? next : getMaximum();
+        return hasNext;
+    }
+
+    public static RangeIterator empty()
+    {
+        return EmptyRangeIterator.instance;
+    }
+
+    public static class EmptyRangeIterator extends RangeIterator
+    {
+        static final RangeIterator instance = new EmptyRangeIterator();
+        EmptyRangeIterator() { super(null, null, 0); }
+        public PrimaryKey computeNext() { return endOfData(); }
+        protected void performSkipTo(PrimaryKey nextToken) { }
+        public void close() { }
+    }
+
+    public static abstract class Builder
+    {
+        public enum IteratorType
+        {
+            CONCAT,
+            UNION,
+            INTERSECTION
+        }
+
+        @VisibleForTesting
+        protected final Statistics statistics;
+
+        @VisibleForTesting
+        protected final PriorityQueue<RangeIterator> ranges;
+
+        public Builder(IteratorType type)
+        {
+            statistics = new Statistics(type);
+            ranges = new PriorityQueue<>(16, Comparator.comparing(RangeIterator::getCurrent));
```

Comment:
We'd have to be pretty deep in the weeds of performance testing to care about this, but do we want to throw a `TODO` in here for the cases where we might know, in advance, the number of ranges we'll add?

Reply:
Does `RangeIntersectionIterator`'s builder ever use this PQ? If not, perhaps it would make sense to move that down into the sub-classes, make `rangeCount()` and `ranges()` abstract, etc. (This base class could still be useful for hiding the statistics, I suppose.)